### PR TITLE
[mlir][Vector] Don't fully unroll transfer_writes of n-D scalable vectors

### PR DESCRIPTION
**Pull request by MacDue**
It is not possible to unroll a scalable vector at compile time. This currently prevents transfer_writes from being lowered to arm_sme.tile_writes (downstream).

### CODE REVIEW

**Comment by MacDue** [comment]
title fix: s/transfer_read/transfer_write/g

File: mlir/test/Conversion/VectorToSCF/vector-to-scf.mlir [hunk 1]
```diff
@@ -737,4 +737,19 @@ func.func @cannot_lower_transfer_read_with_leading_scalable(%arg0: memref<?x4xf3
 // CHECK-SAME:                                                                %[[MEMREF:.*]]: memref<?x4xf32>)
 // CHECK: %{{.*}} = vector.transfer_read %[[MEMREF]][%{{.*}}, %{{.*}}], %{{.*}}, %{{.*}} {in_bounds = [true, true]} : memref<?x4xf32>, vector<[4]x4xf32>
 
+//  -----
 
+// FULL-UNROLL-LABEL: @cannot_fully_unroll_transfer_write_of_nd_scalable_vector
+func.func @cannot_fully_unroll_transfer_write_of_nd_scalable_vector(%arg0: memref<?x?xf32>) {
+  // FULL-UNROLL-NOT: vector.extract {{.*}} : vector<[4]xf32> from vector<[4]x[4]xf32>
+  // FULL-UNROLL-NOT: vector.extract {{.*}} : vector<[4]xi1> from vector<[4]x[4]xi1>
```
Comment (c-rhodes) [inline mlir/test/Conversion/VectorToSCF/vector-to-scf.mlir:745]:
```suggestion
  // FULL-UNROLL-NOT: vector.extract
```
should be sufficient?

File: mlir/test/Conversion/VectorToSCF/vector-to-scf.mlir [hunk 2]
```diff
@@ -737,4 +737,19 @@ func.func @cannot_lower_transfer_read_with_leading_scalable(%arg0: memref<?x4xf3
...
+  // FULL-UNROLL: vector.transfer_write {{.*}} : vector<[4]x[4]xf32>, memref<?x?xf32>
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %v = arith.constant dense<10.0> : vector<[4]x[4]xf32>
+  %dim_a = memref.dim %arg0, %c0 : memref<?x?xf32>
+  %dim_b = memref.dim %arg0, %c1 : memref<?x?xf32>
+  %mask = vector.create_mask %dim_a, %dim_b : vector<[4]x[4]xi1>
+  vector.transfer_write %v, %arg0[%c0, %c0], %mask {in_bounds = [true, true]} : vector<[4]x[4]xf32>, memref<?x?xf32>
```
Comment (c-rhodes) [inline mlir/test/Conversion/VectorToSCF/vector-to-scf.mlir:753]: is the mask necessary? Could this be simplified by removing it or passing it as an arg (likewise for `%v`)?

**Review by c-rhodes — APPROVED** [review]
LGTM cheers

**Review by banach-space — APPROVED** [review]
LGTM, thanks!

**Review by dcaballe — APPROVED** [review]

**Pull request merged**